slsim/Sources/QuasarCatalog/number_density_quasar.py

```diff
@@ -0,0 +1,246 @@
+    :Note: You can adjust these values if needed
+
+    Returns:
+    :rtype: float
+    :return: M_star value.
+    """
+
+    denominator = (np.sqrt(np.exp(xi * z_value)) + np.sqrt(np.exp(xi * z_star))) ** 2
+    if denominator == 0:
+        return np.nan  # or any other suitable value to indicate an error
+    else:
+        return (
+            -20.90
+            + (5 * np.log10(h))
+            - (
+                2.5
+                * np.log10(
+                    np.exp(zeta * z_value) * (1 + np.exp(xi * z_star)) / denominator
+                )
+            )
+        )
+
+
+def dPhi_dM(M, z_value, alpha=-3.31, beta=-1.45, phi_star=5.34e-6 * (0.72**3)):
+    """Calculates dPhi_dM for a given M and redshift according to Eq (10) in Oguri &
+    Marshall (2010).
+
+    Parameters:
+    :param M: Absolute i-band magnitude.
+    :type M: float
+    :param alpha: Bright end slope of quasar luminosity density profile obtained from SDSS DR3 and 2dF surveys (Richards et al. 2005)
+    :param beta: Faint end slope of quasar luminosity density profile obtained from SDSS DR3 and 2dF surveys (Richards et al. 2005)
+    :param phi_star: Function to calculate the renormalization of the quasar luminosity function for a given h.
+
+    Returns:
+    :rtype: float
+    :return: dPhi_dM value.
+    """
+
+    if z_value > 3:
+        # Adjust the bright end slope for redshifts greater than 3 based on observations.
+        alpha_val = -2.58
+    else:
+        # Use the default value for the bright end slope if redshift is 3 or lower.
+        alpha_val = alpha
+
+    M_star_value = M_star(z_value)
+
+    denominator_dphi_dm = (10 ** (0.4 * (alpha_val + 1) * (M - M_star_value))) + (
+        10 ** (0.4 * (beta + 1) * (M - M_star_value))
+    )
+    term1 = phi_star / denominator_dphi_dm
+
+    return term1
+
+
+def cdf_fits_for_redshifts(M_values, random_redshift_values):
+    """Creates Cumulative Distribution Functions (CDF) plots for each randomly generated
+    redshift.
+
+    Parameters:
+    :param M_values: Array of absolute i-band magnitudes.
+    :type M_values: numpy.ndarray
+
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: list
+    :return: List of legend handles for each CDF plot.
+    """
+
+    # Create an empty list to collect artists for the legend
+    legend_handles_cdf = []
+
+    # Plot the CDF vs. M curve for each randomly generated redshift
+    for random_redshift in random_redshift_values:
+        dPhi_dM_random_values = [dPhi_dM(M, random_redshift) for M in M_values]
+        sorted_M_values = np.sort(M_values)
+        cumulative_probabilities = np.cumsum(dPhi_dM_random_values)
+        cumulative_prob_norm = cumulative_probabilities / max(cumulative_probabilities)
+        (line,) = plt.plot(
```

Comment:
avoid having plots mixed into functions that do calculations. Please separate plotting routines from functions defining in and output quantities

Reply:
Created new function 'compute_cdf_data' that handles all numerical calculations needed to generate cdf and inverse cdf curves. Naturally, a new test function was made, called 'test_compute_cdf_data.' It tests for if it outputs a directory, directory keys match with redshift values, cumsum values range between 0 and 1, max cumsum value is 1, M_values are sorted, and M and cumsum arrays are the same length. Ran pytest and it passed.